lint.sh

```diff
@@ -0,0 +1,193 @@
+# - to skip a step, e.g. Skylark linting: "FMT_SKYLINT=false ./lint.sh check".
+#
+# Linters/formatters:
+# - google-java-format: Java code
+# - buildifier: BUILD/WORKSPACE files
+# - skylint: Skylark files (*.bzl) - check only
+# - yapf: Skylark and Python files
+
+BASE="$(pwd)"
+MODE="${1:-check}"
+
+if [ "$MODE" = "check" ]; then
+  JAVA_OPTIONS=--dry-run
+  BUILDIFIER_MODE=check
+  YAPF_OPTIONS=--diff
+else
+  YAPF_OPTIONS=--in-place
+  JAVA_OPTIONS=--replace
+  BUILDIFIER_MODE=fix
+fi
+
+BAZEL_BIN=$(bazel info bazel-bin)
+BAZEL_OUTPUT_BASE=$(bazel info output_base)
+
+function build() {
+  # NOTE: if and when the Skylink target becomes public, use a sh_binary instead
+  # of building everything here?
+  bazel build --color=yes --show_progress_rate_limit=30 \
+    @io_bazel//src/tools/skylark/java/com/google/devtools/skylark/skylint:Skylint \
+    //private:java_format \
+    @com_github_google_yapf//:yapf \
+    @io_bazel_buildifier_linux//file \
+    @io_bazel_buildifier_darwin//file
+}
+
+function format_py_like() {
+  local PATTERN=$1
+  local STYLE=$(cat)
+  local OUTPUT
+
+  OUTPUT=$(find "$BASE" -name "$PATTERN" -exec "$BAZEL_BIN/external/com_github_google_yapf/yapf/yapf" \
+    $YAPF_OPTIONS \
+    "--style=$STYLE" \
+    {} \;)
+  if [ $? != 0 ]; then
+    return 1
+  fi
+  if [ "$MODE" = "check" ] && [ ! -z "$OUTPUT" ]; then
+    echo "$OUTPUT"
+    return 1
+  fi
+}
+
+function format_skylark() {
+  format_py_like "*.bzl" <<'EOF'
+{
+  based_on_style: google,
+  spaces_around_default_or_named_assign: True,
+  blank_lines_around_top_level_definition: 1,
+  indent_width: 2,
+  allow_split_before_dict_value: False,
+  each_dict_entry_on_separate_line: True,
+  split_arguments_when_comma_terminated: True,
+}
+EOF
+}
+
+function format_python() {
+  format_py_like "*.py" <<'EOF'
+{
+  based_on_style: google,
+  spaces_around_default_or_named_assign: True,
+  blank_lines_around_top_level_definition: 2,
+  indent_width: 2,
+  indent_dictionary_value: True
+}
+EOF
+}
+
+function format_bazel() {
+  if [ "$(uname)" = "Darwin" ]; then
+    BUILDIFIER=$BAZEL_OUTPUT_BASE/external/io_bazel_buildifier_darwin/file/downloaded
+  else
+    BUILDIFIER=$BAZEL_OUTPUT_BASE/external/io_bazel_buildifier_linux/file/downloaded
+  fi
+  
+  ERRORS=0
+  $BUILDIFIER -mode=$BUILDIFIER_MODE $(find "$BASE" -name BUILD -type f)
+  ERRORS=$((ERRORS+$?))
+  $BUILDIFIER -mode=$BUILDIFIER_MODE $(find "$BASE" -name WORKSPACE -type f)
+  ERRORS=$((ERRORS+$?))
+
+  # (buildifier cannot format *.bzl files)
+  if [ "$MODE" = "check" ] && ! $BUILDIFIER -mode=check $(find "$BASE" -name "*.bzl" -type f) >/dev/null; then
+    echo "*.bzl BUILDIFIER ERRORS:"
+    for f in $(find "$BASE" -name "*.bzl" -type f); do
+      OUTPUT=$($BUILDIFIER -mode=diff $f)
+      if [ ! -z "$OUTPUT" ]; then
+        echo "$f"
+        echo "$OUTPUT"
+      fi
+    done
+    # That's unfortunate: the only problem is that buildifier does not understand skydocs
```

Comment:
I don’t understand this comment

Reply:
The words are indeed not in the right order :)  I could expand, but the gist of it is that some errors are false positives.